hls_parser.c

```diff
@@ -0,0 +1,192 @@
+/**
+ * HLS frame splitter with PDT parsing functionality
+ * Reference:
+ * - https://blog.csdn.net/leixiaohua1020/article/details/38868499
+ * - https://gist.github.com/tyraeltong/2309baf41e9546b7d757a6477c236418
+ * 
+ * Author: tak-matroid@github.com
+ */
+ 
+#include <stdio.h>
+#include <time.h>
+#include <inttypes.h>
+#include <jpeglib.h> 
+
+#define __STDC_CONSTANT_MACROS
+#define FPS 1.0
+
+#include "libavcodec/avcodec.h"
+#include "libavformat/avformat.h"
+#include "libavutil/imgutils.h"
+#include "libavutil/frame.h"
+#include "libavutil/rational.h"
+#include "libswscale/swscale.h"
+
+#define ASSERT(cond, ...) do { \
+	if (!(cond)) { \
+		printf(__VA_ARGS__); \
+		return -1; \
+	} \
+} while(0)
+
+int process(
+	AVStream *stream,
+	AVCodecContext *inputCodecCtx,
+	AVCodecContext *outputCodecCtx,
+	struct SwsContext *img_convert_ctx,
+	AVPacket *packet, AVFrame *frame, AVFrame *frameYUV)
+{
+	// throttle based on FPS
+	static int64_t last_frame_pts = -1;
+	static double strftime = -1;
+
+	if (av_q2d(stream->time_base) * (frame->pts - last_frame_pts) < (1 / FPS)) {
+		return 0; // SKIP
+	} else {
+		last_frame_pts = frame->pts;
+	} // TODO: this method outputs frames slightly slower than FPS 
```

Comment:
It basically waits for `1 / FPS` before sending the frame to downstream encoder. This no longer matters, as we decide to move forward with `FFmpeg` itself, rather than the `hls_parser.c` script.